### PR TITLE
save logon information in EEPROM

### PR DESCRIPTION
**Pull request by gartonrd**
Kevin - what do you think of this?

### CODE REVIEW

**Comment by gartonrd** [comment]
Doesn't compile. When we unclobbered A3_WebServer, we pulled A3_WebServer from master and I made a few edits. The version of A3_WebServer in master has changes that make it incompatible with the rest of my code because I started with a version which was one merge prior. How best to recover?

**Comment by Multiball** [comment]
I've done my best to reconstruct things here. The code now compiles, and works, as far as I can tell.

The mistake we made was in copy-pasting all the files from a non-repository folder into a repository folder. That totally screwed with the changelogs. One way or another, though, we were going to run into some merge conflicts, because both these branches touched the same files. Oh well; I've resolved those conflicts, and things should be okay.

You won't be able to pull this without an error, because I've rebased on master, which rewrites the commit history. You'll have to delete your local copy of the branch and then check this one out once you've pulled the latest changes. 

I do have a few comments on the approach here; I'll comment on the diff. Let me know if my reconstruction got anything wrong.

File: A5_LogOn.ino [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+      case 0x90:
+        //terminate password string
+        password[LogOnIndex] = '\0';
+        
+        //print Password
+        PrintPassword();
+        
+        //next
+        LogOnAddress += LogOnIndex+1;
+        StartTableAddress = LogOnAddress;
```
Comment (Multiball) [inline A5_LogOn.ino]: I understand the desire to economize space and pack the pattern records as close to the logon records as possible, but I think that ultimately the better is to allocate a constant amount of space for the logon records, and always start the pattern records at the same address. This means that there will usually be some unused space between the end of the logon records and the start of the pattern records, but this prevents the user from having to reupload the pattern information every time the he inputs a new ssid/password.

File: D0_ReadPrint.ino [hunk 2]
```diff
@@ -88,6 +96,60 @@ void WebPrint(String output)
   }
 }
 
+void PrintExecutionStarted()
```
Comment (Multiball) [inline D0_ReadPrint.ino]: Please change these new functions from print functions to string builders, in keeping with the other functions in this file.

**Comment by gartonrd** [comment]
Thanks for fixing things. How would git know if changes have been made by an edit or a copy/paste? I don't understand how this screwed up the change log.

**Comment by gartonrd** [comment]
I agree with your suggestions. 
Do you think 25 characters each is adequate for the ssid and the password?

**Comment by Multiball** [comment]
My password is 23 characters long, and the default password on my router's label is 10 characters long— I guess it's probably enough. If it isn't, we can always bump it.

**Comment by Multiball** [comment]
That's the thing: git doesn't know. If the file is different, it assumes that everything you wrote was a deliberate change. When you pasted D0_ReadPrint, git assumed that you went in and re-typed all the old function definitions and contents. If you had rebased, then the changes in https://github.com/gartonrd/SnowflakesWiFi/pull/5/commits/df97d7cf70c61042d06c3ce83e54f00ddcea4569 wouldn't have clobbered all of the previous changes in https://github.com/gartonrd/SnowflakesWiFi/commit/39e5eef5a0f267816e6bf447705da906d4c37c9d, because that commit would have been applied earlier in your branch's commit history, and it wouldn't have been necessary to manually edit those changes back in (as was done in https://github.com/gartonrd/SnowflakesWiFi/pull/5/commits/f9980484e38e637ecd1c7855c28ed9c7373131eb). In an effort to keep things simple, we inadvertently made more work for ourselves.

**Comment by gartonrd** [comment]
I created and pushed tags for versions 1-5 (V1 - V5). I would like to tag every time we update version number and merge.

**Comment by gartonrd** [comment]
I am curious about something strange that just happened to me. I decided to start fresh by doing a "get clone" in a new folder. The resulting repository did not have the logon and remove-comments branches. But because I am more educated now, I was able to create them by typing "git checkout logon" and "git checkout remove-comments". Before I was educated, you will remember that when I was trying to add my header changes to the remove-comments branch, I did not have the branch, and did not know how to get it. So I did a "git clone" and was then able to navigate to the remove-comments branch. But this time when I did a "git clone" I did not get the remove-comments branch. Why did I get the remove-comments branch last time I did a "git clone" but not this time?

**Comment by Multiball** [comment]
The latest changes you say you pushed don't show up in the list of commits for this branch. Something is missing somewhere— double-check that the changes are committed and pushed.

I don't know why you observed one behavior in one instance and a different behavior in another. My guess is that the behavior you observed was actually the same, but was perceived differently due to differences in expectation that developed between the first and second times. As long as you navigated to the branches with a `git checkout <branch name>`, then you should have gotten those branches in both instances, and it sounds like you did.

**Comment by gartonrd** [comment]
No problems with push - I haven't done that yet, 

You are probably right about the branches. Hard for me to remember exactly what I did.

Got the ssid and password in EEPROM change made. String builder changes tomorrow.

**Comment by gartonrd** [comment]
I understand your desire for consistency, but I am starting to doubt the value of changing all my print functions to string builders. There are some functions at the end of the file that are a mix of print statements and non-print statements. Do you want those changed too?

Everything works. I am afraid of breaking something.

**Comment by gartonrd** [comment]
About what I pushed: I did not push any new commits, but I did push some tags. Adding tags does not require a commit.You should be able to see the tags with the "git tag" command.

**Comment by gartonrd** [comment]
I pushed my ssid and password in EEPROM work. If you are passionate about changing print functions to string builders, go for it. I am not inclined to do so.

**Comment by Multiball** [comment]
Done.

I would not call it a "passion," merely a feeling of responsibility. There's an architectural reason for having string builders: we might not always want to send the strings to Serial. There's currently one other destination for statements we might otherwise print: the browser. Currently, we send only execution records to the browser, but in the future, if we decide to send other things there, we might as well not have to refactor things that we could have done differently in the first place, knowing that that possibility exists. Sending things to the browser was the impetus for changing print functions to string builders in the first place, and it would be nice not to have to perform another such major refactor. We can plan and build for the future.

As you mention, there are plenty of print statements, as in the error functions, that are not currently wrapped in a string builder. I didn't originally touch the error functions because they do things other than print, and I didn't want to mess with that. Ultimately, it would be best for the printing they do to be ripped out and wrapped in a string builder, as would be the case with the WiFi connect messages. Those were already in place, though, and seeing as we're not sending those to the browser, I figure we can change them when it becomes necessary. However, moving forward, it makes sense not to make extra work for ourselves five or however many months down the road, so adding the indirection before we have to will be a net gain.

If something breaks, we fix it. I don't think fear should hold us back from producing quality code. Good testing also helps to allay those fears.

Say, even though we can change the SSID and password, we don't actually disconnect from the previous network and connect to the new network. Is there a decent before/after hook where we could disconnect from the old network and reconnect to the new one?

Also, I should mention: I've never worked with tags before, so you've introduced me to the feature. You are now the expert on git tags.

**Comment by gartonrd** [comment]
The place to disconnect/reconnect the web server after changing the ssid and password would be in A0_ESP, function loop(), case 'L', just after the call to LogOnStateMachine().

What if the ssid/password was bad and the attempt to start the webserver timed out? Make sure a disconnect doesn't bomb in this case.

Of course, you can always reboot after changing the ssid/password to reconnect to the webserver. Two ways to reboot are cycling power and pressing the right button on the update box, 

If you want to add this feature, go for it.

**Comment by gartonrd** [comment]
By the way: I don't feel like an expert on anything about git. But understanding tags is pretty simple. 

There are two kinds of tags, annotated and lightweight. We want to use annotated tags. Here is the procedure for our situation:
1. Find the hash for the commit that merged a pull request. Merging on GitHub will always create such a commit. 
2. Find the most recent version number: git tag
3. Tag the commit: git tag -a V6 -m "put logon information in EEPROM" [commit-hash]
4. Push the tag to GitHub: git push origin V6

You can read all about it here: https://git-scm.com/book/en/v2/Git-Basics-Tagging

**Comment by gartonrd** [comment]
By now you probably realize that marking the code with a version number is important to me. That version number is displayed on the flakes on start up just after the walking pattern. It is really the only way to know what version of the software is executing.

To avoid confusion, I think we need a rule for changing the version number when a branch is created. I think the easiest rule is to set the version number to 0xFFFF. This won't uniquely identify what interim version is executing, but it will identify it as an interim version. When the branch is ready to be merged, the interim 0xFFFF should be changed to one more than the latest, the branch can be merged, and the commit can be tagged. Seems like a lot of work, but when working with a target system it should help avoid confusion.

**Comment by Multiball** [comment]
We can add disconnect/reconnect later, we might as well push this through. We already have graceful failure on failed connect, so we should get that for free when we do disconnect/reconnect.

I'm willing to let you do whatever with the version number, because I don't think it matters unless we have some sort of release process. Version numbers aren't terribly useful during development, because you can always just reupload the code to be sure of which version you're executing. Version numbers are useful when the code is running in the wild, because it lets you know exactly which version of the code the end user you need to help is using. Because we have no release process (which is typically when version numbers are managed), I don't see why it's important. If you're not sure which version of the code you're running, pull latest master and upload that.

But you want a version number, so it's your mandate. I think it's altogether too easy to forget to do something like change a version number first thing when you cut a branch (with a commit where that's the only change), and we'll end up forgetting to do it a lot. Or at least I will, because I've never payed the version number any mind. I'll try, though, if that's what you want.

Anyway, I'm happy with this, I say go ahead and hit the button.

**Comment by gartonrd** [comment]
I want to update Specification.pdf and Software Design.pdf before I press the button. Should get that done tonight.

**Comment by gartonrd** [comment]
Don't pull yet! 
My tag is wrong and I am trying to figure out how to fix it.
I will let you know when I have it fixed.

**Comment by gartonrd** [comment]
Tag is fixed now.
If you have not done a git pull, you can do so now.

If you have done a git pull, you need to do the following
1. git tag -d V6
2. git fetch origin tag V6
3. git show V6 --name-only 
4. verify that commit 85fd86496a6654a7b0247016d3fcb0fd4dc040dc is referenced

The above is the only way to fix the tag if you got it wrong by doing a git pull before the tag was fixed. Doing another git pull will not fix the bad tag.

I am becoming a tag expert, even though I really don't want to.

**Comment by gartonrd** [comment]
Now that this branch is merged, we can move on. 

I suppose the next thing to do is merge the remove-comments branch. I don't care what you do here as long as my headers get changed. So when you are ready, you can merge. I do want you to bump to version 7. I will do the tagging if you want me to.

The next feature I would like to see is a way to synchronize snowflake nodes. Some time ago I started two nodes at the same time by pressing the reset buttons on both. After a day, they were about 1/2 second out of sync. So we need a way to re-sync them periodically.

Not knowing how possible this is, I would like to propose the following. Display a drop down in the browser that lets the user select how often he would like to re-sync. Offer these choices: Never, 60 minutes, 30 minutes, 15 minutes. At the designated time (on the quarter hour, half hour, or whole hour), send a StopExecution, then a StartExecution.

I am open to other ways to accomplish this. But I will be leaving for the cabin tomorrow, and won't be available to discuss again until Monday. Let me know what you think.

**Comment by Multiball** [comment]
Ah, now we are beginning to enter into the wonderful world of distributed systems.

In order for all nodes to be synchronized, we need to be able to talk to every node. The browser is directly connected to only one node at a time: the node whose IP address you connect to. This means that the browser can't talk directly to every node: only one at a time.

Well, sort of. The browser can talk to any node whose IP address it knows. If any node can provide the IP address of every node, then the browser can get the list of IP addresses from any node it connects to. Then, when it's time to sync, it can send a "stop/restart" signal (or perhaps even better, a "stop, then restart at a specific time" signal) to every node in the list.

If that's the approach we want to take, then there needs to be some sort mechanism for nodes to discover each other. A node can't know a priori the IP address of all other nearby nodes; as you've observed, IP addresses change, and if you're unplugging and replugging nodes trying to figure out how exactly you want to wrangle them, they can change at a moment's notice. To have a robust solution, hard-wiring IP addresses isn't an option, and would probably rely on router configuration in order to ensure that it actually works reliably. To have a robust solution, nodes need to be able to discover each other independently.

This should be possible, though I'm not clear on the details right now. A node would have to somehow broadcast its IP address, so that other nodes, scanning for such IP addresses (this scanning is the detail that I'm fuzziest on), would be able to send it its own IP address, and the node would be able to thus keep a list of all IP addresses so encountered. Then, when the browser schedules a stop/restart event, the node can broadcast that event to all the IPs it knows about, and they can propagate it to all the nodes whose IPs it knows, and so on, and so on.

That's one model, highly distributed. Another model, which would require at least different code to run on some boxes than on others, and at most would require different hardware, is to have one master control unit whose job is only to manage the other nodes. All nodes connect to this master node, and it tracks all of them. This would probably be the node that the browser connects to, as well. All stop/restart events are scheduled by this node and broadcast to all nodes connected to it.

These are the two models that occur to me off the top of my head. If the nodes have an internal clock, then we might be able to set them to automatically stop/restart on certain intervals, and they'd never have to talk to each other. Configuring this value dynamically would still require some sort communication between nodes, though. What do you think?

**Comment by gartonrd** [comment]
I have two thoughts:
1. Can't we just open a browser window for each node? Then the router would know about all of the nodes. Initially, we want to synchronize only two nodes - the one outside and one inside to mirror what is happening outside. I don't know if we will ever have a separate node for each snowflake, ie 16 nodes to synchronize. But maybe. Still, 16 browser windows isn't impossible.
2. I don't think we can rely on an internal clock in each node for synchronizing. If nodes could accurately keep time, my experiment starting two of them at the same time would have been more successful. But maybe the nodes could reference a clock on some website somewhere. Windows says it synchronizes to time.windows.com. And there is time.gov.

Thoughts?

**Comment by Multiball** [comment]
1. You can open a browser window for each node, but pressing the button on each window simultaneously is not humanly possible. But now that I think about it, we might be able to try... multicast. I believe that you can give an ESP a multicast address, and that might be just what we want to broadcast a signal to many nodes at once. Now, I just have to figure out how multicast works to find out if this is actuallya solution to this problem or not.
2. Hmm, hom, yes. If there's no internal clock, then it's not useful to try to say "restart at this epoch timestamp," which was my idea. Time syncing is typically best left to the professionals, and I'd be hesitant to try it ourselves, but if that's the only option, we can give it a shot.

**Comment by gartonrd** [comment]
1A. I wasn't thinking the user would have to press a button in each window, only that by opening a window for each node the router could gain access to each node. The actual sync would be done by the router automatically sending a command to each node.
1B. The multicast idea sounds like it is worth pursuing
2. If we can get access to an external clock, I think the rest is easy. The problems are: 1) Internal clocks aren't synchronized, 2) Internal clocks don't retain accuracy. If able to read an external clock, an internal clock could be created which would be accurate enough to read the external clock close to sync time and get the exact moment to sync.
3. Could all nodes get the time from the computer, rather than a website. We don't need an absolute time at each node, only the same time at every node.

**Comment by Multiball** [comment]
1A. What "router?" The router that assigns IP addresses to everything on the LAN? That router does nothing but route. You can't execute arbitrary code on it. (Well, you probably could, but you'd have to hack it something fierce.)
1B. I'll look into it. There's still the problem of knowing the IP address of every node.
2-3. This is theoretically possible, but might be difficult, because you'd have to set up some sort of server on the computer that will respond to time requests from the flake nodes. This would still probably be easier than pulling from an external web site.

**Comment by gartonrd** [comment]
2-3. We could keep an internal clock that is sort of accurate. When it gets close to sync time, we could send a request for the precise time from the computer, update our internal clock, and do a start/stop at sync time.Not sure if this is possible. Since we are a server, can we initiate requests?

**Comment by Multiball** [comment]
As long as you know the IP address of the node you want to talk to, you can make a request, sure. There's nothing to stop a server from being a client, too.

**Comment by gartonrd** [comment]
Here is a tag team proposal:
1. You provide a way for me to request a reading from the PC clock.
2. You make that reading available to me.
3. The reading is a 32 bit time stamp.
4. I will use this mechanism to do the synchronization:
4.1 StartExecution()
      Get time from PC
      Start internal clock which will be updated every 10 ms.
4.2 HandleTimerIRQ()
      Add 10 ms to internal clock
      If((sync time) - (internal clock time) < 2000ms)
         StopExecution()
         Execute = 2
      End if
4.3 loop()
      If(Execute == 2)
         Get time from PC
         delayMicroseconds((sync time) - (PC time))
         StartExecution();
      End if

If you understand the above, you are welcome to do the whole thing

**Comment by Multiball** [comment]
Are you sure that you want to make the flake node execution dependent on code that's executed on a CPU that's not a flake node? If you have to run some sort of flake time daemon on an external box, then the IP address of that box needs to be known and given to each flake node somehow, and the user has to make sure the flake time daemon is running every time he wants to run the flakes. That's drastically increasing the burden on the end user, which was something we've been trying to avoid.

**Comment by gartonrd** [comment]
I am not knowledgeable enough in this area to know what the best approach is. What is your best recommendation?

**Comment by Multiball** [comment]
Well, I'm not knowledgeable enough in this area to know what the best approach is, either. I was just throwing out ideas.

Architecturally, I think we should make it a goal to avoid forcing the user to run additional software on a box other than the node boxes. It's also true that I just think it would be really cool to properly distribute things. This leaves us with the hub-and-spoke model, where we have one specialized flake node running different code than the others, and it's the only one that the user interacts with; and the completely distributed model, where the nodes keep track of each other's IP addresses and the user can connect to any one of them and thence communicate with all of them. 

The hub-and-spoke model is probably the easiest to implement. In this model, there are two types of node: and "execution node," that runs flake patterns; and a "master node," that tracks all the other flake nodes and resets them when necessary. The master node is the one  that the user communicates directly with via a web browser. The user doesn't communicate with the execution nodes directly, but they instead receive all their commands (including reset commands) through the master node. 

In the other model, there is only one kind of node, execution nodes. All nodes execute flake instructions. They find each other (I'm not certain of the exact mechanism for this) and communicate their addresses to each other. Any node the user connects to via the web browser relays the instructions it receives to the other nodes that it has the IP address of.

The more I think about it, the more I favor the hub-and-spoke model for practicality and reliability. If the goal is synchronization, the hub-and-spoke model will be able to trigger a reset on all nodes in quicker succession than the distributed model. Not having a clock is a pain. If we force the user to connect to the master node (say, to start execution), then we can get the system time from the browser and relay it to the master node, which can propagate it to the execution nodes. I'm not convinced that running our own clock is reliable, though. Can we be certain that the period for updating the clock is consistent? The interrupt-driven nature of ESP execution would seem to preclude this.

The time it takes to go down the list might matter, though.

**Comment by gartonrd** [comment]
I thought reading a clock from a website or the PC would be an easy task. Apparently not. 

So, let's try another approach: Optionally sync each time a node gets to the end of the run table. I say optionally, because I think the user should be able to configure whether the node waits for a sync message to wrap to the start of the table, sends sync signal(s) to other node(s) when it reaches the end of the table, or doesn't sync at all. I can see how to do all of this except sending and receiving the sync message. If you could provide utilities to do these two functions, I can do the rest. Don't worry about trying to automatically discover other nodes which might be involved in the sync process. For starters we will make the user enter the IP addresses of these nodes. It might be nice if the user could do this in the browser window. The interface would need to allow the user to set: no sync, wait for message before wrapping, send sync messages to a list of IP addresses. This list would be entered manually. If it is is too hard to do this in the browser, I can provide this interface over the USB port. 

I think it would be best if you did as much as you could, and then let me finish it up. But I could do my part first and call stubbed routines to send/receive the sync message. 

Let me know how you would like to proceed.

**Comment by Multiball** [comment]
What exactly is the mechanism for synching?

It sounds like what you're proposing is for the user to send a list of IP addresses to each node, and then, when a node reads the last record of a run table (record `0x91`), send a "table end" message to each IP address that the user has provided. When a node has recieved a "table end" message from each IP address the user has provided, it then proceeds to the next iteration of the run table.

Is this accurate? This sounds like "send sync messages to a list of IP addresses." What is "wait for message before wrapping?" Which message? Who is sending this message?

**Comment by gartonrd** [comment]
Sorry for the lapse in communication. I want to get the house painted this summer and we have had a string of nice days, so I have been out painting.

Your description of my proposal is accurate. I don't understand your questions. The "wait for message before wrapping" refers to slave nodes that must receive a sync message from a master node before proceeding to the start of the table when reaching the end of table. 

What may not be clear, is that I can conceive of several arrangements given this one mechanism. There could be one or more master nodes that send sync messages to one or more slave nodes. The master nodes may or may not also be slave nodes.

Is this doable?

**Comment by Multiball** [comment]
The description I gave of your proposal has no master or slave nodes; all nodes are equal. All nodes have the same list of IP addresses; they all wait until they have received a "table end" message from every node on the list (besides themselves). Do you want to introduce a master/slave model? It sounds like that's what you had in mind.

**Comment by gartonrd** [comment]
I have been working on the AC board, and I have a few changes to make to accommodate it. Might be a while until I get a break from the painting project. Just giving you a heads up.

Trip to Champaign and Indianapolis planned for mid June. Hope to have the changes done by then. Will only take me a day or two, but who knows when the weather will turn bad and I can't work on the painting.

**Comment by gartonrd** [comment]
In your latest description, every node is both a master and a slave. Master = sends sync messages when reach end of table. Slave = wait for sync message before proceeding to start of table when reaching end of table.

The mechanism may be configured this way, or several other ways. For example, only one node might be configured as a master and the rest as slaves.

If all nodes have the capability to be a master, a slave, or both. we have flexibility to experiment with different arrangements.

**Pull request merged**